File: web_app/tests/business/GameControllerTest.py
"""
GameControllerTest.py
"""
import unittest
from ..persistence.DataAccessStub import DataAccessStub
from web_app.application.Services import Services
from web_app.business.GameController import GameController

class GameControllerTest(unittest.TestCase):
    """Unit tests for the Question class"""

    @classmethod
    def setUpClass(cls):
        """Call once"""
        Services.close_data_access()
        Services.create_data_access(\
            altDataAccessService=DataAccessStub("application"))

    @classmethod
    def tearDownClass(cls):
        """Call once"""
        Services.close_data_access()

    def setUp(self):
        """Call before every test case"""
        self.max_questions = 3
        self.dummy_game_controller = GameController.get_instance(self.max_questions)

    def tearDown(self):
        """Call after every test case"""
        GameController.destroy()

    def test_singleton(self):
        """Make sure GameController is a Singleton"""
        print "Testing GameController: Singleton"
        first = GameController.get_instance(self.max_questions)
        second = GameController.get_instance(self.max_questions)
        self.assertEquals(first, second)

    def test_start(self):
        """Make sure the start function works"""
        print "Testing GameController: Start"
        self.assertFalse(self.dummy_game_controller.is_started)
        self.dummy_game_controller.start()
        self.assertTrue(self.dummy_game_controller.is_started)

    def test_increase_score(self):
        """Make sure increase_score works"""
        print "Testing GameController: increase_score"
        self.dummy_game_controller.start()
        self.assertEquals(0, self.dummy_game_controller.score)
        self.dummy_game_controller.increase_score()
        self.assertEquals(1, self.dummy_game_controller.score)

File: android_app/app/src/test/java/comp4350/triviasmack/tests/business/GameControllerTest.java
package comp4350.triviasmack.tests.business;

import junit.framework.TestCase;

import java.util.ArrayList;

import comp4350.triviasmack.application.Main;
import comp4350.triviasmack.application.Services;
import comp4350.triviasmack.business.GameController;
import comp4350.triviasmack.objects.Question;

public class GameControllerTest extends TestCase {

    private GameController dummyGameController;

    public GameControllerTest(String arg0) {
        super(arg0);
    }

    @Override
    protected void setUp() throws Exception {
        Services.closeServerAccess();
        Services.createServerAccess(new ServerAccessStub(Main.numQuestions));
        dummyGameController = GameController.getInstance();
    }

    @Override
    protected void tearDown() throws Exception {
        Services.closeServerAccess();
        GameController.destroy();
    }

    public void testSingleton(){
        System.out.println("Testing GameController: Singleton");
        GameController first = GameController.getInstance();
        GameController second = GameController.getInstance();
        assertEquals(first, second);
    }

    public void testStart(){
        System.out.println("Testing GameController: Start");
        assertFalse(dummyGameController.isStarted());
        dummyGameController.start();
        assertTrue(dummyGameController.isStarted());
    }

    public void testIncreaseScore(){
        System.out.println("Testing GameController: IncreaseScore");
        dummyGameController.start();
        assertEquals(0, dummyGameController.getScore());
        dummyGameController.increaseScore();
        assertEquals(1, dummyGameController.getScore());
    }

    public void testAccessors(){
        System.out.println("Testing GameController: Accessors");

        dummyGameController.start();

        ArrayList<Question> questions = new ArrayList<>();

        dummyGameController.getQuestions(questions);

        assertNotNull(questions);

        for (int i = 0; i < questions.size(); i++){
            assertTrue(questions.get(i) instanceof Question);
        }
    }

    public void testGetNextQuestions(){
        System.out.println("Testing GameController: getNextQuestion");

        dummyGameController.start();

        Question questionObj = dummyGameController.getNextQuestion();

        assertNotNull(questionObj);
    }
}
